src/layers/domain/core/questionnaire.py

```diff
@@ -27,49 +30,238 @@ class Question(BaseModel):
+        for question_name, response in self.responses.items():
+            is_valid, failed_rules = self.validate_question_response_rules(
+                question_name, response
+            )
+
+            if not is_valid:
+                invalid_responses_rules.append((question_name, response, failed_rules))
+
+        return invalid_responses_rules
+
+    def validate_question_response_choices(self, question_name, response):
+        question = next(
+            (q for q in self.questionnaire.questions if q.name == question_name), {}
+        )
+        valid_choices = question.choices
+        if isinstance(response, list):
+            for item in response:
+                # Check if the question has choices and if each response is one of the valid choices
+                if valid_choices and item not in valid_choices:
+                    return False, valid_choices
+
+        elif valid_choices and response not in valid_choices:
+            return False, valid_choices
+        return True, None
+
+    def validate_questionnaire_responses_choices(self):
+        invalid_responses_choices = []
+        for question_name, response in self.responses.items():
+            is_valid, valid_choices = self.validate_question_response_choices(
+                question_name, response
+            )
+
+            if not is_valid:
+                invalid_responses_choices.append(
+                    (question_name, response, valid_choices)
+                )
+
+        return invalid_responses_choices
+
+    def validate_responses_correspond_to_questionnaire(self):
+        # Check if questionnaire responses correspond to the questions in the questionnaire
+        for question_name in self.responses:
+            if question_name not in [q.name for q in self.questionnaire.questions]:
+                raise InvalidResponseError(
+                    f"Unexpected answer for the question '{question_name}'. The questionnaire '{self.questionnaire.name}' does not contain this question."
+                )
+        return True
+
+    def validate_multiple_responses_allowed(self):
+        invalid_responses = []
+        # check if multiple responses gievn when not expected
+        for q in self.questionnaire.questions:
+            if q.multiple == False and isinstance(
+                self.responses[q.name], list
+            ):  # lenght of list >1? Single response as a list?
+                invalid_responses.append((q.name, self.responses[q.name]))
+
+        if len(invalid_responses) > 0:
+            invalid_responses_given = []
+            for q.name, self.responses[q.name] in invalid_responses:
+                invalid_response = f"Question '{q.name}' does not allow multiple responses. Response given: {self.responses[q.name]}."
+                invalid_responses_given.append(invalid_response)
+
+            raise InvalidResponseError(invalid_responses_given)
+
+        return True
+
+    def validate_and_get_invalid_responses(self):
```

Comment:
so this is the function where the `@validator("responses", each_item=True)` will go, and then since we can then let pydantic do the error handling for us, simplify the entire function (with the other suggestions) to

```python
    @validator("responses", each_item=True))
    def validate_responses(self, response, values: dict[str, Questionnaire]):
        questionnaire = values.get("questionnaire")
        question_name, answers = response

        question = questionnaire.get(question_name)
        if question is None:
                raise InvalidQuestionResponse("question doesnt exist")    

        if not question.multiple and len(answers) > 1:
                raise InvalidQuestionResponse("too many")    

        for answer in answers:
            if not isinstance(answer, question.type):
                raise InvalidQuestionResponse("bad type")

            for validation_rule in validation_rules:
                validation_rule(answer)  # should raise a value error

            if question.choices and answer not in question.choices:
                raise InvalidQuestionResponse("bad choice")
```